### ISSUE TITLE
Apm Agend doesn't support async/await?

### ISSUE THREAD

**Issue by thrcorrea** [elastic/apm-agent-nodejs#873]
**Describe the bug**
When a define a route wich uses knex to connect do the database, and use async await at the function definition, the apm agent does not send the sql span, not showing the sql info at the request timeline at kibana. The apm agent does not support async/await

**To Reproduce**
1. Use this config '...'
2. Then call '....'
3. Then do '....'

**Expected behavior**
To show the sql span at the request timeline.

**Environment (please complete the following information)**
- OS: Linux
- Node.js version: v11.10.0
- APM Server version: 6.6.1
- Agent version: 2.5.1

**How are you starting the agent? (please tick one of the boxes)**
- [X] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [ ] Starting node with `-r elastic-apm-node/start`

**Additional context**

- Agent config options 
  <details>
    <summary>Click to expand</summary>

    ```
    serviceName: 'service',
    ```
  </details>
- `package.json` dependencies:
  <details>
    <summary>Click to expand</summary>

    ```
    "elastic-apm-node": "^2.5.1",
    "express": "^4.16.4",
    "knex": "^0.16.3",
    "mysql2": "^1.6.5"
    ```
  </details>

**Comment by watson** [comment]
Does the database span it self show up - just without the SQL query - or is the entire database span missing from the transaction?

Would you mind sharing a code snippet of how you're making the query?

**Comment by thrcorrea** [comment]
The entire database span is missing. I created a project with and example:

https://github.com/thrcorrea/apm-test

**Comment by watson** [comment]
Thanks for sharing the test app. I've been able to reproduce the problem now. I don't think it's directly async/await related, but you're right that this is what triggers it in this case. Other apps with async/await do work though. So I guess it's something with the combination of how knex operates. I'll investigate further...

**Comment by watson** [comment]
Update: It's somehow related to a combination of async/await and async hooks. A temporary workaround is to set `asyncHooks: false` in the agent config.

I'll continue to investigate to find a more permanent solution...

**Comment by watson** [comment]
This issue might be the same as discussed in #897

**Comment by MartinKolarik** [comment]
Same problem here, knex + async/await + asyncHooks (+ something else? not sure) = no active transaction for mysql spans. I'm on node 12. Turning off `asyncHooks` solves this issue but as I understand, it might break context propagation for some other cases?

**Comment by jsumners** [comment]
We are seeing this problem. Screenshot 1 here is with `ELASTIC_APM_ASYNC_HOOKS=true`:

![:visit_id - Vivaldi 2019-10-07 14-27-07](https://user-images.githubusercontent.com/321201/66338237-b7901d80-e90e-11e9-969b-4e558680291e.png)

Screenshot 2 is the same transaction path with `ELASTIC_APM_ASYNC_HOOKS=false`:

![:visit_id - Vivaldi 2019-10-07 14-29-09](https://user-images.githubusercontent.com/321201/66338303-de4e5400-e90e-11e9-903d-d0c6e820860d.png)

This is just a series of SQL statements being executed in order via `await db.query`. They are not executed within a database transaction.

Note: in the first screenshot there is a `SET` statement happening from the `mysql2` module's `onConnect` hook to set the timezone for the session.

**Comment by PaulMougel** [comment]
FWI, we have the same issue when we use [Objection.js](https://vincit.github.io/objection.js/) (an SQL ORM): queries are not properly associated with the current transaction.

The trace logs show that no transaction is found, even though in our case we are processing a route response from hapi. If we manually create a span it is attached to the correct transaction, and knex requests are properly instrumented and associated with the transaction.

What is the recommended course of action?

**Comment by tgriesser** [comment]
I've been subscribed to this issue because I was curious if there was a solution (proper tracing in node is a difficult thing to do). So this is mostly just a complete guess here, but I wonder if this is because knex now uses a module called "tarn" as its pooling library:

https://github.com/vincit/tarn.js/

It looks like there's a instrumentation for generic-pool, which knex previously used:

https://github.com/elastic/apm-agent-nodejs/blob/master/lib/instrumentation/modules/generic-pool.js

Perhaps adding a similar shim for tarn would help associate the trace?

**Comment by Qard** [comment]
It's because the promise hooks feature in V8, which Node.js relies on for async_hooks, does not work with Thenables, which both Knex and Objection.js are using. This is unfixable without investing some time in fixing the issue in V8. See: https://github.com/nodejs/node/issues/22360

**Comment by brandoncollins7** [comment]
Having this issue as well now, which led me to this thread. Are there any workarounds?

**Comment by JCMais** [comment]
The issue mentioned above states that Node.js v14.5.0 has the thenable / async_hook compat issue fixed, I cannot test right now, but if anyone is willing to test and report back it would be appreciated.

**Comment by Qard** [comment]
Yep, it's fixed in 14.6.x+. It hasn't been backported to 12.x yet, but shouldn't be too hard to backport.

**Comment by JCMais** [comment]
Can confirm that this issue is fixed when running the apm agent on Node.js >= 14.6
